Archive/Imo/Imo1982Q3.lean

```diff
@@ -0,0 +1,96 @@
+/-
+Copyright (c) 2024 Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Violeta Hernández Palacios, Alex Brodbelt
+-/
+import Mathlib.Algebra.Order.BigOperators.Ring.Finset
+import Mathlib.Algebra.GeomSum
+import Mathlib.Data.NNReal.Basic
+
+/-!
+# IMO 1982 Q3
+
+Consider infinite sequences $\{x_n\}$ of positive reals such that $x_0 = 0$ and
+$x_0 \ge x_1 \ge x_2 \ge \ldots$
+
+a) Prove that for every such sequence there is an $n \ge 1$ such that:
+
+$$\frac{x_0^2}{x_1} + \ldots + \frac{x_{n-1}^2}{x_n} \ge 3.999$$
+
+b) Find such a sequence such that for all $n$:
+
+$$\frac{x_0^2}{x_1} + \ldots + \frac{x_{n-1}^2}{x_n} < 4$$
+
+The solution is based on Solution 1 from the
+[Art of Problem Solving](https://artofproblemsolving.com/wiki/index.php/1982_IMO_Problems/Problem_3)
+website. For part a, we use Sedrakyan's lemma to show the sum is bounded below by
+$\frac{4n}{n + 1}$, which can be made arbitrarily close to $4$ by taking large $n$. For part b, we
+show the sequence $x_n = 2^{-n}$ satisfies the desired inequality.
+-/
+
+open Finset NNReal
+
+variable {x : ℕ → ℝ} {n : ℕ} (hn : n ≠ 0) (hx : Antitone x)
+
+namespace Imo1982Q3
+include hn hx
+
+/-- `x (n + 1)` is at most the average of `x 1`, `x 2`, ..., `x n`. -/
+lemma ineq₁ : ∑ k ∈ range (n + 1), x k ≤ (∑ k ∈ range n, x k) * (1 + 1 / n) := by
+  rw [sum_range_succ, mul_one_add, add_le_add_iff_left, mul_one_div,
+    le_div_iff₀ (mod_cast hn.bot_lt), mul_comm, ← nsmul_eq_mul]
+  conv_lhs => rw [← card_range n, ← sum_const]
+  refine sum_le_sum fun k hk ↦ hx (le_of_lt ?_)
+  simpa using hk
+
+variable (h0 : x 0 = 1) (hp : ∀ k, 0 < x k)
+include h0 hp
+
+/-- We combine Sedrakyan and the previous inequality. -/
+lemma ineq₂ : (∑ k ∈ range n, x (k + 1) + 1) ^ 2 / ((∑ k ∈ range n, x (k + 1)) * (1 + 1 / n))
+    ≤ ∑ k ∈ range (n + 1), x k ^ 2 / x (k + 1) := by
+  apply le_trans _ (sq_sum_div_le_sum_sq_div _ x (fun k _ ↦ hp (k + 1)))
+  gcongr
+  · exact sum_pos (fun k _ ↦ hp _) nonempty_range_succ
+  · exact add_nonneg (sum_nonneg fun k _ ↦ (hp _).le) zero_le_one
+  · rw [sum_range_succ', h0]
+  · exact ineq₁ hn (hx.comp_monotone @Nat.succ_le_succ)
+
+/-- We move `1 + 1 / n` out of the fraction. -/
+lemma ineq₃ : (∑ k ∈ range n, x (k + 1) + 1) ^ 2 / (∑ k ∈ range n, x (k + 1)) * n / (n + 1)
+    ≤ ∑ k ∈ range (n + 1), x k ^ 2 / x (k + 1) := by
+  convert ineq₂ hn hx h0 hp using 1
+  field_simp
+
+/-- Finally, AM-GM gives us the main lower bound for the RHS. -/
+lemma ineq₄ : 4 * n / (n + 1) ≤ ∑ k ∈ range (n + 1), x k ^ 2 / x (k + 1) := by
+  apply le_trans _ (ineq₃ hn hx h0 hp)
+  gcongr
+  rw [le_div_iff₀]
+  · simpa using four_mul_le_sq_add (∑ k ∈ range n, x (k + 1)) 1
+  · exact sum_pos (fun k _ ↦ hp _) (nonempty_range_iff.2 hn)
```

Comment:
Can you please turn this into a single `calc` block showcasing the powerful combination of `calc` + `gcongr` + `field_simp` + `ring` + `linarith` + `omega`?

Reply:
I've merged `ineq₂` through `ineq₄` into a single calc block. I found it cleaner to keep `ineq₁` as an auxiliary lemma, since I instantiate it with `x ⚬ succ` rather than just `x`.